### PR TITLE
buffered_operate_to! for BigFloat dot product

### LINKED ISSUE #189: How is one supposed to compute the dot product of BigFloat vectors, utilizing mutability?
I'm trying to optimize the Tulip linear optimizer for the `BigFloat` arithmetic. I noticed that in multiple places in Tulip's source a weighted sum of multiple dot products is computed, so I made a function `dot_weighted_sum_inner` that uses MA. The weights are actually always either `1` or `-1`.

The trouble is, `MA.add_dot` again uses the regular `BigFloat` multiplication and the generic `LinearAlgebra.dot`! See allocation profiling flamegraph screenshot:

tulip_dot_allocs

In the end this results in a slight *increase* in allocation!

How am I supposed to use MA to do this?

### PR DESCRIPTION
**Pull request by nsajko**
Fixes #189

### CODE REVIEW

**Comment by nsajko** [comment]
Depends on #192

**Comment by nsajko** [comment]
I'm looking into making a more meaningful accuracy test, which shouldn't fail arbitrarily.

**Comment by nsajko** [comment]
FTR this is ready for review.

File: test/bigfloat_dot.jl [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+function dot_test_relative_error(x::V, y::V) where {V<:AbstractVector{BigFloat}}
+    buf = MA.buffer_for(LinearAlgebra.dot, V, V)
+
+    input = (x, y)
+    backup = map(backup_bigfloats, input)
+
+    output = BigFloat()
+
+    MA.buffered_operate_to!!(buf, output, LinearAlgebra.dot, input...)
```
Comment (blegat) [inline test/bigfloat_dot.jl:33]: Does this allocate ? If not, you can use what is in `test/utils.Jl` to test that it does not allocate

**Comment by nsajko** [comment]
ping

File: test/bigfloat_dot.jl [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+function ulper(size::Int, bias::Real, iter_cnt::Int)
+    let s = size, b = bias, c = iter_cnt
+        () -> max_rand_dot_ulps(s, b, c)
+    end
+end
+
+@testset "prec:$prec size:$size bias:$bias" for (prec, size, bias) in
+                                                Iterators.product(
+    (32, 64),
+    (10000, 100000),
```
Comment (blegat) [inline test/bigfloat_dot.jl:75]: I'm wondering why we need such size for testing

Comment (nsajko) [reply test/bigfloat_dot.jl]: I'm not an expert, so the comments I added might make someone who knows better cringe, but I hope it's mostly OK. I based the code and my understanding of the subject matter on the Wikipedia page on Kahan summation and Higham's 1993 paper:

> Higham, Nicholas J. "The accuracy of floating point summation." SIAM Journal on Scientific Computing 14.4 (1993): 783-799.

**Comment by nsajko** [comment]
The test failure on Julia nightly doesn't seem related to this PR.

File: test/bigfloat_dot.jl (hunk 2 above)
Comment (blegat) [reply test/bigfloat_dot.jl]: Thanks for adding the comments, could you add the ref to Higham's paper as well then ? The algorithm is a bit cryptic without a reference :grin:

**Review by blegat — APPROVED** [review]
Thanks!

**Pull request merged**